File: ruleMatch/src/main/scala/com/bigdata/rulematch/scala/service/impl/HBaseQueryServiceImpl.scala
package com.bigdata.rulematch.scala.service.impl

import com.bigdata.rulematch.scala.conf.EventRuleConstant
import org.apache.commons.lang3.StringUtils
import org.apache.hadoop.hbase.TableName
import org.apache.hadoop.hbase.client.{Connection, Get}
import org.apache.hadoop.hbase.util.Bytes
import org.slf4j.{Logger, LoggerFactory}

/**
 * 查询hbase的服务实现类
 */
class HBaseQueryServiceImpl {
  private val logger: Logger = LoggerFactory.getLogger(this.getClass.getName.stripSuffix("$"))

  private var hbaseConn: Connection = null

  def this(hbaseConn: Connection) = {
    this()
    this.hbaseConn = hbaseConn
  }

  /**
   * 根据给定的rowkey，查询hbase，判断查询出来的标签值与规则中的标签值是否一致
   * @param rowkey
   * @param userProfileConditions
   * @return
   */
  def userProfileConditionIsMatch(rowkey: String,
                                  userProfileConditions: Map[String, String]) = {
    var isMatch = true

    //获取到hbase的表
    val table = hbaseConn.getTable(TableName.valueOf(EventRuleConstant.HBASE_USER_PROFILE_TABLE_NAME))

    val get = new Get(Bytes.toBytes(rowkey))
    val family = "f"

    //需要查询的用户标签,只需要查询规则条件中的标签即可,不需要全部查询
    val tags = userProfileConditions.keySet
    tags.foreach(tag => {
      get.addColumn(Bytes.toBytes(family), Bytes.toBytes(tag))
    })

    // 执行get查询
    val result = table.get(get)

    //判断查询出来的标签值与用户画像规则条件中的值是否一致
    //这里做了简化,正常情况下,Map的value可以使用一个二元组,第一个元素是比较符,第二个元素是比较的值
    //这样就会很灵活，比如判断年龄是否大于18岁,目前这种简化的方式，只能判断等于18岁
    val tagsIterator = tags.iterator
    while (tagsIterator.hasNext && isMatch) {
      val tag = tagsIterator.next()
      val valueBytes: Array[Byte] = result.getValue(Bytes.toBytes(family), Bytes.toBytes(tag))

      val valueStr = Bytes.toString(valueBytes)

      val userProfileConditionsValue = userProfileConditions.getOrElse(tag, "")

      if (!StringUtils.equals(valueStr, userProfileConditionsValue)) {
        isMatch = false
        logger.debug(s"用户画像类规则不匹配, 标签名: ${tag}, 规则中要求的值: ${userProfileConditionsValue}, " +
          s"hbase中查询到的值: ${valueStr}")
      }
    }

    isMatch
  }
}

File: ruleMatch/src/main/java/com/bigdata/rulematch/java/service/impl/HBaseQueryServiceImpl.java
package com.bigdata.rulematch.java.service.impl;

import com.bigdata.rulematch.java.conf.EventRuleConstant;
import org.apache.commons.lang.StringUtils;
import org.apache.hadoop.hbase.TableName;
import org.apache.hadoop.hbase.client.Connection;
import org.apache.hadoop.hbase.client.Get;
import org.apache.hadoop.hbase.client.Result;
import org.apache.hadoop.hbase.client.Table;
import org.apache.hadoop.hbase.util.Bytes;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.Map;
import java.util.Set;

/**
 * 查询hbase的服务实现类
 *
 * @author HanDongfang
 * @version 1.0
 * @create 2021-12-21  14:17
 */
public class HBaseQueryServiceImpl {
    private Logger logger = LoggerFactory.getLogger(this.getClass().getName());

    private Connection hbaseConn = null;

    /**
     * 初始化连接
     *
     * @param hbaseConn
     */
    public HBaseQueryServiceImpl(Connection hbaseConn) {
        this.hbaseConn = hbaseConn;
    }

    /**
     * 根据给定的rowkey，查询hbase，判断查询出来的标签值与规则中的标签值是否一致
     *
     * @param rowkey
     * @param userProfileConditions
     * @return
     */
    public boolean userProfileConditionIsMatch(String rowkey, Map<String, String> userProfileConditions) {
        boolean isMatch = true;

        //获取到hbase的表
        Table table = null;
        try {
            table = hbaseConn.getTable(TableName.valueOf(EventRuleConstant.HBASE_USER_PROFILE_TABLE_NAME));
            //生成get
            Get get = new Get(Bytes.toBytes(rowkey));
            //列族
            String family = "f";
            Set<String> keySet = userProfileConditions.keySet();
            //添加查询列
            for (String s : keySet) {
                get.addColumn(Bytes.toBytes(family), Bytes.toBytes(s));
            }

            // 执行get查询
            Result result = table.get(get);

            for (Map.Entry<String, String> userProfileEntry : userProfileConditions.entrySet()) {
                String key = userProfileEntry.getKey();
                byte[] value = result.getValue(Bytes.toBytes(family), Bytes.toBytes(key));
                String valueStr = value.toString();
                String userProfileEntryValue = userProfileEntry.getValue();
                //如果与查询值不匹配则停止比较并返回false
                if (!StringUtils.equals(valueStr, userProfileEntryValue)) {
                    isMatch = false;
                    break;
                }
            }

            return isMatch;
        } catch (IOException e) {

            logger.error(String.format("获取hbase表数据出错: %s", e));
            return false;
        }
    }
}
